### PR TITLE
Add LLaMA decoder and causal LM

### PR DESCRIPTION
**Pull request by danieldk**
## Description

LLaMA is pretty similar to NeoX, with some small modifications:

- Gated Linear Units
- RMS norm rather than layer norm
- No parallel attention

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/feedforward.py [hunk 1]
```diff
@@ -1,30 +1,51 @@
-from typing import Optional
+from typing import Callable, Dict, Optional
 
 import torch
 from torch import Tensor
 from torch.nn import Linear, Module
 
 from .activations import GeluFast, GeluNew
 
+_ACTIVATIONS: Dict[str, Callable[[], Module]] = {
```
Comment (shadeMe) [inline curated_transformers/models/feedforward.py:9]: Can't we use `Type` instead of `Callable[...]`?

File: curated_transformers/models/llama/config.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        num_attention_heads: int = 32,
+        rotary_base=10000,
+        rotary_fraction=0.25,
+    ):
+        """
+        :param dropout_prob: Dropout to apply after attention.
+        :param hidden_width: Hidden width of the transformer.
+        :param num_attention_heads: Number of attention heads.
+        :param rotary_base: Base in signifying the rotary embedding period.
+        :param rotary_fraction: fraction of hidden width to apply rotary
```
Comment (shadeMe) [inline curated_transformers/models/llama/config.py:31]:
```suggestion
        :param rotary_fraction: Fraction of hidden width to apply rotary
```

File: curated_transformers/models/llama/decoder.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        :param positions: Input positions. Positions are needed to
+            look up rotary embeddings. Normally, these positions are calculated
+            automatically. But if the positions deviate for some reason, they
+            can be provided through this argument.
+        :param store_cache: Whether to cache the key/value representations for
+            future reuse.
+        :returns: Decoder representations of the given pieces.
+
+        Shapes:
+            input_ids, attention_mask, positions - (batch, seq_len)
```
Comment (shadeMe) [inline curated_transformers/models/llama/decoder.py:75]: Shapes as part of param docstrings.

File: curated_transformers/models/normalization.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        Apply RMS normalization to a tensor.
+
+        :param input:
+            The tensor to apply normalization to.
+        :returns:
+            Normalized tensor.
+        """
+        # Zhang & Sennrich, Equation 4
+        rms = (
+            input.to(torch.float32)
```
Comment (shadeMe) [inline curated_transformers/models/normalization.py:37]: Maybe a comment to clarify why we're doing the calculation in f32?

File: curated_transformers/models/feedforward.py (hunk 1 above)
Comment (danieldk) [reply curated_transformers/models/feedforward.py]: Sounds good, fixed.

File: curated_transformers/models/llama/config.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+
+@dataclass
+class LLaMAAttentionConfig:
+    """LLaMA (Touvron et al., 2023) attention configuration."""
+
+    dropout_prob: float
+    hidden_width: int
+    num_attention_heads: int
+    qkv_mode: QkvMode
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/llama/config.py]: Removed, also in NeoX, where this was originally copied from.

File: curated_transformers/models/llama/decoder.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/models/llama/decoder.py]: Fixed, also in NeoX, where this was originally copied from.

File: curated_transformers/models/normalization.py (hunk 4 above)
Comment (danieldk) [reply curated_transformers/models/normalization.py:39]: Added.
